pennylane_lightning/src/util/Cache.hpp

```diff
@@ -0,0 +1,196 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Least Recently Updated (LRU) cache policy class implementation.
+ */
+
+#pragma once
+
+/* ======================================================= */
+/* LRU Library Header */
+/* ======================================================= */
+
+#include <list>
+#include <unordered_map>
+#include <vector>
+
+#include <limits>
+#include <utility>
+
+namespace Pennylane {
+
+namespace Util {
+
+// Boost implementation of a hash combine:
+// https://www.boost.org/doc/libs/1_35_0/doc/html/boost/hash_combine_id241013.html
+struct hash_function {
+    std::size_t
+    operator()(const std::pair<std::vector<size_t>, size_t> &key) const {
+        std::size_t combined_hash_value = 0;
+
+        for (auto &term : key.first) {
+            combined_hash_value ^= std::hash<size_t>()(term) + 0x9e3779b9 +
+                                   (combined_hash_value << 6) +
+                                   (combined_hash_value >> 2);
+        };
+        combined_hash_value ^= std::hash<size_t>()(key.second) + 0x9e3779b9 +
+                               (combined_hash_value << 6) +
+                               (combined_hash_value >> 2);
+        return combined_hash_value;
+    }
+};
```

Comment:
I wonder since the hash will be dependent upon the datatype being cached, if this should be housed elsewhere, such as within the source file using the cache. Ideally, we want to be as generalisable as possible with this, and having a hash function for a single key type alongside the cache utility may be somewhat restrictive.
Having the user of the cache define the necessary hash where it is used may be a better idea. Otherwise, if this file can be appropriately templated to work with any (reasonable) type, this could also work.

What do you think?

Reply:
I think you are right. I will move it to the StateVector source code and adapt the class accordingly.